service/matching/tasklist/task_list_manager_test.go

```diff
@@ -667,32 +667,119 @@ func TestTaskListManagerGetTaskBatch(t *testing.T) {
 		task.Finish(nil)
 	}
 	assert.Equal(t, taskCount-rangeSize, tm.GetTaskCount(taskListID))
-	tasks, _, isReadBatchDone, err = tlm.taskReader.getTaskBatch()
+	tasks, _, isReadBatchDone, err = tlm.taskReader.getTaskBatch(tlm.taskAckManager.GetReadLevel(), tlm.taskWriter.GetMaxReadLevel())
 	assert.NoError(t, err)
 	assert.True(t, 0 < len(tasks) && len(tasks) <= rangeSize)
 	assert.True(t, isReadBatchDone)
 	tlm.Stop()
 }
 
+func TestTaskListReaderPumpAdvancesAckLevelAfterEmptyReads(t *testing.T) {
+	const taskCount = 5
+	const rangeSize = 10
+	const nLeaseRenewals = 15
+
+	controller := gomock.NewController(t)
+	mockPartitioner := partition.NewMockPartitioner(controller)
+	mockPartitioner.EXPECT().GetIsolationGroupByDomainID(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return("", nil).AnyTimes()
+	mockDomainCache := cache.NewMockDomainCache(controller)
+	mockDomainCache.EXPECT().GetDomainByID(gomock.Any()).Return(cache.CreateDomainCacheEntry("domainName"), nil).AnyTimes()
+	mockDomainCache.EXPECT().GetDomainName(gomock.Any()).Return("domainName", nil).AnyTimes()
+
+	logger := testlogger.New(t)
+	timeSource := clock.NewRealTimeSource()
+	tm := NewTestTaskManager(t, logger, timeSource)
+	taskListID := NewTestTaskListID(t, "domainId", "tl", 0)
+	cfg := defaultTestConfig()
+	cfg.RangeSize = rangeSize
+
+	tlMgr, err := NewManager(
+		mockDomainCache,
+		logger,
+		metrics.NewClient(tally.NoopScope, metrics.Matching),
+		tm,
+		cluster.GetTestClusterMetadata(true),
+		mockPartitioner,
+		nil,
+		func(Manager) {},
+		taskListID,
+		types.TaskListKindNormal.Ptr(),
+		cfg,
+		timeSource,
+		timeSource.Now(),
+	)
+	require.NoError(t, err)
+	tlm := tlMgr.(*taskListManagerImpl)
+
+	// simulate lease renewal multiple times without writing any tasks
+	for i := 0; i < nLeaseRenewals; i++ {
+		tlm.taskWriter.renewLeaseWithRetry()
+	}
+
+	err = tlm.Start() // this call will also renew lease
+	require.NoError(t, err)
+	defer tlm.Stop()
+
+	// we expect AckLevel to advance
+	expectedAckLevel := int64(rangeSize) * nLeaseRenewals
+
+	// wait until all tasks are read by the task pump and enqeued into the in-memory buffer
+	// at the end of this step, ackManager readLevel will also be equal to the buffer size
+	assert.True(t, awaitCondition(func() bool {
+		return tlm.taskAckManager.GetAckLevel() == expectedAckLevel
+	}, time.Second))
```

Comment:
you are using real time source. is 1s sufficient here to let the pump reach to this expected level? just asking in case this can be flaky at times

Reply:
I'll change it to 5.
We use the same 1s in other tests for similar purpose, but I think there is no need since we'll break as soon as success anyway.